src/Raven.Client/Util/TcpUtils.cs

```diff
@@ -241,51 +256,91 @@ private static TcpClient NewTcpClient(TimeSpan? timeout, bool useIPv6)
             return tcpClient;
         }
 
-        internal static async Task<(TcpClient Client, string Url)> ConnectAsyncWithPriority(
-            TcpConnectionInfo info,
-            TimeSpan? tcpConnectionTimeout
+        public delegate Task<TcpConnectionHeaderMessage.SupportedFeatures> NegotiationCallback(string url, TcpConnectionInfo tcpInfo, Stream stream, JsonOperationContext context, List<string> logs = null);
+
+        internal static async Task<(TcpClient Client, Stream Stream, string ChosenUrl, TcpConnectionHeaderMessage.SupportedFeatures supportedFeatures)> ConnectSecuredTcpSocket(
+            TcpConnectionInfo info, 
+            X509Certificate2 cert,
+#if !(NETSTANDARD2_0 || NETSTANDARD2_1 || NETCOREAPP2_1)
+            CipherSuitesPolicy cipherSuitesPolicy,
+#endif  
+            TcpConnectionHeaderMessage.OperationTypes operationType,
+            NegotiationCallback negotiationCallback,
+            JsonOperationContext negContext,
+            TimeSpan? timeout,
+            List<string> logs = null
 #if TCP_CLIENT_CANCELLATIONTOKEN_SUPPORT
             ,
             CancellationToken token = default
 #endif
             )
         {
-            TcpClient tcpClient;
+            TcpClient tcpClient = null;
+            Stream stream = null;
+            TcpConnectionHeaderMessage.SupportedFeatures supportedFeatures = null;
 
-            if (info.Urls != null)
+            string[]  infoUrls = info.Urls == null ? new string[] {} : info.Urls.Append(info.Url).ToArray();
```

Comment:
maybe better to use a hashset here with ordinal ignore case comparer? Do we care about the order?

Reply:
info.Url should be last